keras_nlp/samplers/top_p_sampler.py

```diff
@@ -0,0 +1,173 @@
+from keras_nlp.samplers.sampler import base_sampler_args_docstring
+from keras_nlp.samplers.sampler import call_args_docstring
+from keras_nlp.samplers.sampler import sample_args_docstring
+from keras_nlp.utils.python_utils import format_docstring
+
+
+@format_docstring(
+    base_sampler_args=base_sampler_args_docstring, call_args=call_args_docstring
+)
+@keras.utils.register_keras_serializable(package="keras_nlp")
+class TopPSampler(Sampler):
+    """Top-P Sampler class.
+    This sampler implements top-p search algorithm. Top-p search selects tokens
+    from the smallest subset of output probabilities that sum to greater than
+    `p`. Put in another way, top-p will first order token predictions by
+    likelihood, and ignore all tokens after the cumulative probability of
+    selected tokens exceeds `p`, then select a token from the remaining tokens.
+
+    Args:
+        p: float, the `p` value of top-p.
+        seed: int, defaults to None. The random seed.
+        {{base_sampler_args}}
+
+    Call Args:
+        {{call_args}}
+
+    Examples:
+    ```python
+    BATCH_SIZE = 8
+    VOCAB_SIZE = 10
+    FEATURE_SIZE = 16
+    START_ID = 1
+
+    # Create a dummy model to predict the next token.
+    model = keras.Sequential(
+        [
+            keras.Input(shape=[None]),
+            keras.layers.Embedding(
+                input_dim=VOCAB_SIZE,
+                output_dim=FEATURE_SIZE,
+            ),
+            keras.layers.Dense(VOCAB_SIZE, activation="softmax"),
+        ]
+    )
+
+    # Define a function that outputs the next token's probability for each token
+    # in the input sequence.
+    def token_probability_fn(inputs, mask):
+        return model(inputs)
+
+    prompt = tf.fill((BATCH_SIZE, 1), START_ID)
+
+    sampler = keras_nlp.samplers.TopPSampler(p=0.1)
+    # Print the generated sequence (token ids).
+    print(sampler(prompt, token_probability_fn, 10))
+    ```
+    """
+
+    def __init__(
+        self,
+        p=0.1,
+        seed=None,
+        jit_compile=True,
+        run_eagerly=False,
+    ):
+        self.p = p
+        self.seed = seed
+        super().__init__(jit_compile, run_eagerly)
+
+    @format_docstring(sample_args=sample_args_docstring)
+    def sample(
+        self, prompt, token_probability_fn, mask, num_steps, from_logits=True
+    ):
+        """Sampling logic implementation.
+
+        Args:
+            {{sample_args}}
+        """
+        batch_size, max_length = tf.shape(prompt)[0], tf.shape(prompt)[1]
+        max_length = tf.cast(max_length, num_steps.dtype)
+        length = max_length - num_steps
+
+        def one_step(length, prompt, mask):
+            probs = token_probability_fn(prompt, mask)
+            pred = tf.gather(
+                probs, tf.repeat(length - 1, batch_size), axis=1, batch_dims=1
+            )
+            if from_logits:
+                pred = keras.activations.softmax(pred, axis=-1)
+            # Sort preds in descending order.
```

Comment:
The main change I could see us making if we want to, is splitting the overload points into `sample` and `sample_step` (I think some discussion of this on other PR?).

Overload `sample` if you want to control the whole process (e.g. beam search). Overload `sample_step` if you want to control simply going from one probability distribution to one sample. E.g. the body of this class could look like:

```python
def sample_step(preds):
     sorted_preds, sorted_indices = tf.math.top_k(
          pred, k=tf.shape(preds)[1], sorted=True
      )
      # Calculate cumulative probability distribution.
      cumulative_probs = tf.math.cumsum(sorted_preds, axis=-1)
      # Create a mask for the tokens to keep.
      keep_mask = cumulative_probs <= self.p
      # Shift to include the last token that exceed p.
      shifted_keep_mask = tf.concat(
          [tf.ones_like(keep_mask[:, :1]), keep_mask[:, :-1]], axis=-1
      )
      # Filter out unmasked tokens and sample from filtered distribution.
      probs = tf.where(
          shifted_keep_mask,
          sorted_preds,
          tf.zeros(tf.shape(pred), dtype=sorted_preds.dtype),
      )
      sorted_next_token = tf.random.categorical(
          tf.math.log(probs), 1, seed=self.seed
      )
      return tf.gather_nd(
          sorted_indices, sorted_next_token, batch_dims=1
      )
```

This might improve the readability of our simple samples, while still keeping full extensibility.

Reply:
As long as we can make `BeamSampler` an outlier, I am down to this refactoring!

Actually we can move one step further to only leave `get_next_token` open, which takes in a prob distribution and returns the next token, and the rest updating logic can be shared across those samplers. 

Reflected this change in the PR.

Reply:
Yes, this is what I was hoping for! 🚀